src/settings/project/mod.rs

```diff
@@ -84,54 +84,50 @@ impl Manifest {
-        if release && self.workers_dot_dev.is_some() {
-            failure::bail!(
-                "The --release flag is not compatible with use of the workers_dot_dev field"
-            )
-        }
-
-        if release {
-            message::warn("--release will be deprecated");
-        }
-
-        let environment = match environment_name {
-            Some(environment_name) => match &self.env {
-                Some(environment_table) => {
-                    let environment = environment_table.get(environment_name);
-                    match environment {
-                        Some(environment) => Some(environment),
-                        None => failure::bail!(format!(
-                            "{} Could not find environment with name {}",
-                            emoji::WARN,
-                            environment_name
-                        )),
-                    }
+    ) -> Result<Option<&Environment>, failure::Error> {
+        if let Some(environment_name) = environment_name {
+            if let Some(environment_table) = &self.env {
+                if let Some(environment) = environment_table.get(environment_name) {
+                    Ok(Some(environment))
+                } else {
+                    failure::bail!(format!(
+                        "{} Could not find environment with name {}",
+                        emoji::WARN,
+                        environment_name
+                    ))
                 }
-                None => failure::bail!(format!(
+            } else {
+                failure::bail!(format!(
                     "{} There are no environments specified in your wrangler.toml",
                     emoji::WARN
-                )),
-            },
-            None => None,
-        };
+                ))
+            }
+        } else {
+            Ok(None)
+        }
+    }
 
+    // TODO: when --release is deprecated, this will be much easier
+    fn negotiate_with_terrorists(
+        &self,
+        environment: Option<&Environment>,
+        release: bool,
+    ) -> Result<(Option<String>, bool), failure::Error> {
         let deprecate_warning =
             "Please specify the workers_dot_dev boolean in the top level of your wrangler.toml";
         let wdd_failure = format!(
-            "{} Your environment should only include `workers_dot_dev` or `route`",
+            "{} Your environment should only include `workers_dot_dev` or `route`. If you are trying to publish to workers.dev, remove `route` from your wrangler.toml, if you are trying to publish to your own domain, remove `workers_dot_dev`.",
             emoji::WARN
         );
 
         // TODO: deprecate --release, remove warnings and parsing
         // switch wrangler publish behavior to act the same at top level
         // and environments
         // brace yourself, this is hairy
-        let workers_dot_dev = match environment {
+        let workers_dot_dev: bool = match environment {
```

Comment:
Part of me wonders if there is a way to further simplify the logic around the truth table for `workers_dot_dev`, `release`, and `route`. What we have here is totally serviceable--I think, however, this logic could become even more understandable for folks if we reduced the nested matches and focused on easy-to-follow compound conditionals.

For example, we could use logic that could do the following pseudocode:
1. Extract a specified environment only once using one match statement.
2. If environment does not exist (is `None`, so top level configs used):
    a. If `release == true` and `workers_dot_dev.is_some()`, throw error message stating conflict between using `--release` and `workers_dot_dev`.
    b. If `release == true` and `workers_dot_dev.is_none()`, warn of deprecation and return false (proceed to deploy to route)
    c. If `release == false` and `self.workers_dot_dev.is_some()` and `!self.route.is_empty()`, bail with `wdd_failure`
    d. If `release == false` and `self.workers_dot_dev.is_some()` and `self.route.is_empty()`, return `workers_dot_dev` value
    e. If `release == false` and `self.workers_dot_dev.is_none()` return true.

Reply:
I'll work on this today :)

Reply:
Actually, can we leave this so we can merge refactor today? @gabbifish

Reply:
Makes sense to me! I'll save the link to this convo and we can revisit it after this is merged.